internal/indexer/oci.go

```diff
@@ -0,0 +1,201 @@
+
+// WithIndexRef configures the charts index OCI reference instead of letting the library
+// using the default host/index:latest one.
+//
+// 	opt := WithIndexRef("my.oci.domain/index:prod")
+//
+func WithIndexRef(r string) OciIndexerOpt {
+	return func(opts *ociIndexerOpts) {
+		opts.reference = r
+	}
+}
+
+// WithBasicAuth configures basic authentication for the OCI host
+//
+// 	opt := WithBasicAuth("user", "pass")
+//
+func WithBasicAuth(user, pass string) OciIndexerOpt {
+	return func(opts *ociIndexerOpts) {
+		opts.username = user
+		opts.password = pass
+	}
+}
+
+// WithInsecure configures insecure connection
+//
+// 	opt := WithInsecure()
+//
+func WithInsecure() OciIndexerOpt {
+	return func(opts *ociIndexerOpts) {
+		opts.insecure = true
+	}
+}
+
+// WithHost configures the OCI host
+//
+// 	opt := WithHost("my.oci.domain")
+//
+func WithHost(h string) OciIndexerOpt {
+	return func(opts *ociIndexerOpts) {
+		opts.url = h
+	}
+}
+
+// ociIndexer is an OCI-based Indexer
+type ociIndexer struct {
+	reference string
+	resolver  remotes.Resolver
+}
+
+// DefaultIndexName is the name for the OCI artifact with the index
+const DefaultIndexName = "index"
+
+// DefaultIndexTag is the tag for the OCI artifact with the index
+const DefaultIndexTag = "latest"
+
+// NewOciIndexer returns a new OCI-based indexer
+func NewOciIndexer(opts ...OciIndexerOpt) (Indexer, error) {
+	opt := &ociIndexerOpts{}
+	for _, o := range opts {
+		o(opt)
+	}
+
+	u, err := url.Parse(opt.url)
+	if err != nil {
+		return nil, errors.Wrapf(ErrInvalidArgument, "invalid OCI host URL: %+v", err)
+	}
+	resolver := newDockerResolver(u, opt.username, opt.password, opt.insecure)
+
+	ind := &ociIndexer{
+		reference: opt.reference,
+		resolver:  resolver,
+	}
+	if ind.reference == "" {
+		host := strings.Trim(strings.Join([]string{u.Host, u.Path}, "/"), "/")
+		ind.reference = fmt.Sprintf("%s/%s:%s", host, DefaultIndexName, DefaultIndexTag)
+		klog.Infof("Remote index was not provided. Using default location %q", ind.reference)
+	}
+
+	return ind, nil
+}
+
+// VacAssetIndexLayerMediaType is a media type used in VAC to store a JSON containing the index of
+// charts and containers in a repository
+const VacAssetIndexLayerMediaType = "application/vnd.vmware.tac.asset-index.layer.v1.json"
+// VacAssetIndexConfigMediaType is a media type used in VAC for the configuration of the layer above
+const VacAssetIndexConfigMediaType = "application/vnd.vmware.tac.index.config.v1+json"
+
+// DefaultIndexFilename is the default filename used by the library to upload the index
+const DefaultIndexFilename = "asset-index.json"
+
+// Get implements Indexer
+func (ind *ociIndexer) Get(ctx context.Context) (idx *api.Index, e error) {
+	// Allocate folder for temporary downloads
+	dir, err := os.MkdirTemp("", "indexer")
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to create temporary indexer directory")
+	}
+	defer func() {
+		err := os.RemoveAll(dir)
+		if e == nil {
```

Comment:
why do you have this `e` error variable?

Reply:
When a deferred function returns an error as part of the signature, we should always handle that case. This is pretty common with file closers (`defer f.Close`). A common pattern to handle the error of a deferred function so we can return handle the error outside is to use named return variables (`e error` in this case) so we can check the returned value (returned error) inside the deferred function body.

```
err := os.RemoveAll(dir)
		if e == nil {
```

Hence, here we are removing all directories (which might raise errors), then checking if the function has exited with errors (you have to decided whether to append or not the deferred function error here) and, if not, replace the returned error with your new error (or `nil` error, if everything went fine).

I learned this at https://www.joeshaw.org/dont-defer-close-on-writable-files/ so you may find it interesting too.